posts/2018-06-15-typedapi.md

```diff
@@ -0,0 +1,329 @@
+```
+
+#### Make the request
+We have all the data we need to make an IO request but nothing to execute it. We change that now. By adding an HTTP backend. But we don't want to expose this implementation detail through our code. What we want is a generic description of a request action and that sounds again like a job for type classes.
+
+```scala
+trait ApiRequest[M, F[_], C, Out] {
+
+  def apply(data: (Uri, Queries), client: C): F[Out]
+}
+```
+
+We have to specialize that for the set of methods we have:
+
+```scala
+trait GetRequest[C, F[_], Out] extends ApiRequest[GetCall, C, F, Out]
+
+...
+```
+
+Let's say we want http4s as our backend. Then we just have to implement these `traits` using http4s functionality.
+
+#### Make it a whole
+We have a bunch of type classes which in theory do a request, but so far they are completely useless. To make a working piece of code out of it we have to connect them.
+
+```scala
+def derive[Api <: HList, El <: HList, KIn <: HList, VIn <: HList, M, Out, F[_], C]
+  (api: ApiTypeCarrier[Api], client: C)
+  (implicit fold: Lazy[TypeLevelFoldLeft.Aux[Api, Fold], (El, KIn, VIn, M, Out)]
+            builder: RequestBuilder[El, KIn, VIn],
+            request: ApiRequest[M, F, C, Out]): VIn => F[Out] = vin => request(builder.apply(vin, List.newBuilder, Map.empty), client)
+```
+
+This first approach gives us the desired function but has a major drawback. You have to fix `F[_]` somehow and the only way is to set it explicitly. But by doing that you are forced to provide definitions for all the type parameters. Furthermore, this function isn't really convenient. To use it you have to create and pass an `HList` and as we said before, we don't want to expose something like that.
+
+To fix the first problem we simply add a helper class which moves the step of defining the higher kind `F[_]` to a separate function call:
+
+```scala
+final class ExecutableDerivation[El <: HList, KIn <: HList, VIn <: HList, M, O](builder: RequestDataBuilder[El, KIn, VIn], input: VIn) {
+
+  final class Derivation[F[_]] {
+
+    def apply[C](client: C)(implicit req: ApiRequest[M, C, F, O]): F[O] = {
+      val data = builder(input, List.newBuilder, Map.empty, Map.empty)
+
+      req(data, cm)
+    }
+  }
+
+  def run[F[_]]: Derivation[F] = new Derivation[F]
+}
+```
+
+Making a function of arity `Length[VIn]` out of `Vin => F[O]`is possible by using `shapeless.ops.function.FnFromProduct`.
+
+When we apply both solutions we end up with:
+
+```scala
+def derive[H <: HList, Fold, El <: HList, KIn <: HList, VIn <: HList, M, Out]
+  (apiList: ApiTypeCarrier[H])
+  (implicit fold: Lazy[TypeLevelFoldLeft.Aux[H, Nothing, (El, KIn, VIn, M, Out)]],
+            builder: RequestDataBuilder[El, KIn, VIn],
+            vinToFn: FnFromProduct[VIn => ExecutableDerivation[El, KIn, VIn, M, Out]]): vinToFn.Out = 
+  vinToFn.apply(input => new ExecutableDerivation[El, KIn, VIn, M, Out](builder, input))
+```
```

Comment:
If I was the author of this post, I would probably slip a joke there about how much bigger the signature of this function is compared to its body, which is exactly what we're looking for when doing typelevel programming :)

Reply:
tuff task ... me finding a good joke 😄 . I will try to come up with one.